lib/internal/util/parse_args/parse_args.js

```diff
@@ -0,0 +1,315 @@
+'use strict';
+
+const {
+  ArrayPrototypeForEach,
+  ArrayPrototypePushApply,
+  ArrayPrototypeShift,
+  ArrayPrototypeSlice,
+  ArrayPrototypePush,
+  ArrayPrototypeUnshiftApply,
+  ObjectPrototypeHasOwnProperty: ObjectHasOwn,
+  ObjectEntries,
+  StringPrototypeCharAt,
+  StringPrototypeIncludes,
+  StringPrototypeIndexOf,
+  StringPrototypeSlice,
+} = primordials;
+
+const {
+  validateArray,
+  validateObject,
+  validateString,
+  validateUnion,
+  validateBoolean,
+} = require('internal/validators');
+
+const {
+  findLongOptionForShort,
+  isLoneLongOption,
+  isLoneShortOption,
+  isLongOptionAndValue,
+  isOptionValue,
+  isOptionLikeValue,
+  isShortOptionAndValue,
+  isShortOptionGroup,
+  objectGetOwn,
+  optionsGetOwn,
+} = require('internal/util/parse_args/utils');
+
+const {
+  codes: {
+    ERR_INVALID_ARG_VALUE,
+    ERR_PARSE_ARGS_INVALID_OPTION_VALUE,
+    ERR_PARSE_ARGS_UNKNOWN_OPTION,
+    ERR_PARSE_ARGS_UNEXPECTED_POSITIONAL,
+  },
+} = require('internal/errors');
+
+function getMainArgs() {
+  // This function is a placeholder for proposed process.mainArgs.
+  // Work out where to slice process.argv for user supplied arguments.
+
+  // Electron is an interesting example, with workarounds implemented in
+  // Commander and Yargs. Hopefully Electron would support process.mainArgs
+  // itself and render this workaround moot.
+  //
+  // In a bundled Electron app, the user CLI args directly
+  // follow executable. (No special processing required for unbundled.)
+  // 1) process.versions.electron is either set by electron, or undefined
+  //    see: https://www.electronjs.org/docs/latest/api/process#processdefaultapp-readonly
+  // 2) process.defaultApp is undefined in a bundled Electron app, and set
+  //    in an unbundled Electron app
+  //    see: https://www.electronjs.org/docs/latest/api/process#processdefaultapp-readonly
+  // (Not included in tests as hopefully temporary example.)
+  /* c8 ignore next 3 */
+  if (process.versions?.electron && !process.defaultApp) {
+    return ArrayPrototypeSlice(process.argv, 1);
+  }
+
+  // Check node options for scenarios where user CLI args follow executable.
+  const execArgv = process.execArgv;
+  if (StringPrototypeIncludes(execArgv, '-e') ||
```

Comment:
what happens with all these checks when the executable filename itself contains these strings?

Reply:
1) `process.execArgv` does not include the filename

2) Actually, `StringPrototypeIncludes` should probably be `ArrayPrototypeIncludes` for at least readability if not functionality. (That might have been an inaccurate Primordial conversion from the original code, but as the unit tests are passing it seems to be working as intended!)

Reply:
ok, then what if there’s an argument like `'foo-e'` :-p

Reply:
> ok, then what if there’s an argument like 'foo-e' :-p

`proces.execArgv` is an array, so the use of `StringPrototypeIncludes` was a bit confusing. I've switched to using `ArrayPrototypeIncludes`, and I've confirmed that it requires an exact match, so that `foo-e` would not match.

Reply:
aha, then everything's fine :-) thanks